ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala

```diff
@@ -0,0 +1,214 @@
+    : Future[Either[ValidationFailed, TransformedSubmission[T]]] =
+    Envelope.open(envelope) match {
+      case Right(Envelope.SubmissionMessage(submission)) =>
+        val declaredInputs = submission.getInputDamlStateList.asScala
+        ledgerStateAccess.inTransaction { stateOperations =>
+          for {
+            readStateInputs <- Future.sequence(
+              declaredInputs.map(
+                key =>
+                  stateOperations
+                    .readState(keyToBytes(key))
+                    .map { stateValue =>
+                      key -> stateValue.map(bytesToStateValue)
+                  }
+              )
+            )
+            damlLogEntryId = allocateLogEntryId()
+            readInputs: Map[DamlStateKey, Option[DamlStateValue]] = readStateInputs.toMap
+            missingInputs = declaredInputs -- readInputs.filter {
+              case (_, value) => value.isDefined
+            }.keySet
+          } yield {
+            if (missingInputs.nonEmpty) {
+              Left(MissingInputState(missingInputs.map(keyToBytes)))
+            } else {
+              Try {
+                val (logEntry, damlStateUpdates) =
+                  processSubmission(damlLogEntryId, recordTime, submission, readInputs)
+                postProcessResult(
+                  damlLogEntryId,
+                  flattenInputStates(readInputs),
+                  (logEntry, damlStateUpdates),
+                  stateOperations)
+              }.fold(
+                exception => Left(ValidationError(exception.getLocalizedMessage)),
+                result => Right(TransformedSubmission(result)))
+            }
+          }
+        }
+      case _ =>
+        Future.successful(
+          Left(ValidationError(s"Failed to parse submission, correlationId=$correlationId")))
+    }
+
+  private def flattenInputStates(
+      inputs: Map[DamlStateKey, Option[DamlStateValue]]): Map[DamlStateKey, DamlStateValue] =
+    inputs.collect {
+      case (key, Some(value)) => key -> value
+    }
+}
+
+object SubmissionValidator {
+  type RawBytes = Array[Byte]
+  type RawKeyValuePairs = Seq[(RawBytes, RawBytes)]
+
+  type StateMap = Map[DamlStateKey, DamlStateValue]
+  type LogEntryAndState = (DamlLogEntry, StateMap)
+
+  def create(
+      ledgerStateAccess: LedgerStateAccess,
+      allocateNextLogEntryId: () => DamlLogEntryId = () => allocateRandomLogEntryId())(
+      implicit executionContext: ExecutionContext): SubmissionValidator = {
+    val participantId: ParticipantId =
+      ParticipantId.assertFromString(ledgerStateAccess.participantId)
+    new SubmissionValidator(
+      ledgerStateAccess,
+      processSubmission(participantId),
+      allocateNextLogEntryId)
+  }
+
+  def allocateRandomLogEntryId(): DamlLogEntryId =
+    DamlLogEntryId.newBuilder
+      .setEntryId(ByteString.copyFromUtf8(UUID.randomUUID().toString))
+      .build()
+
+  private lazy val engine = Engine()
```

Comment:
We might want to run multiple validators in one process, e.g. for testing purposes. Right now sharing the engines is not a big issue as they only share the compiled packages, but later on we’ll surely add metrics etc. which shouldn’t be shared. Consider not making the engine a singleton. Otoh, we can refactor when we need to.

Reply:
That's a good point. Let's leave this for a later PR if that's fine with you -- i.e., we may want to add a `processSubmission` wrapper that works with multiple `Engine` instances running on separate threads.

Reply:
SGTM